packages/miniflare/src/workers/kv/namespace.worker.ts

```diff
@@ -0,0 +1,275 @@
+import assert from "node:assert";
+import { Buffer } from "node:buffer";
+import {
+  DELETE,
+  DeferredPromise,
+  GET,
+  HttpError,
+  KeyValueStorage,
+  MiniflareDurableObject,
+  PUT,
+  RouteHandler,
+  maybeApply,
+} from "miniflare:shared";
+import { KVHeaders, KVLimits, KVParams } from "./constants";
+import {
+  decodeKey,
+  decodeListOptions,
+  validateGetOptions,
+  validateKey,
+  validateListOptions,
+} from "./validator.worker";
+
+interface KVParams {
+  key: string;
+}
+
+// Returns value as an integer or undefined if it isn't one
+function normaliseInt(value: string | number | undefined): number | undefined {
+  switch (typeof value) {
+    case "string":
+      return parseInt(value);
+    case "number":
+      return Math.round(value);
+  }
+}
+
+function createMaxValueSizeError(length: number, maxValueSize: number) {
+  return new HttpError(
+    413,
+    `Value length of ${length} exceeds limit of ${maxValueSize}.`
+  );
+}
+class MaxLengthStream extends TransformStream<Uint8Array, Uint8Array> {
+  readonly signal: AbortSignal;
+  readonly length: Promise<number>;
+
+  constructor(maxLength: number) {
+    const abortController = new AbortController();
+    const lengthPromise = new DeferredPromise<number>();
+
+    let length = 0;
+    super({
+      transform(chunk, controller) {
+        length += chunk.byteLength;
+        // If we exceeded the maximum length, don't enqueue the chunk, but don't
+        // error the stream, so we get the correct final length in the error
+        if (length <= maxLength) {
+          controller.enqueue(chunk);
```

Comment:
The chunk that pushes us over the limit will not be enqueued. Did you consider enqueuing a part of that chunk so the stream is exactly maxLength?